packages/grid/x-data-grid/src/hooks/features/rows/useGridRows.ts

```diff
@@ -331,6 +332,69 @@ export const useGridRows = (
     [apiRef, logger],
   );
 
+  const replaceRows = React.useCallback<GridRowApi['replaceRows']>(
+    (firstRowToRender, lastRowToRender, newRows) => {
+      if (props.signature === GridSignature.DataGrid) {
+        throw new Error(
+          [
+            "MUI: You can't replace rows using `apiRef.current.replaceRows` on the DataGrid.",
+            'You need to upgrade to DataGridPro or DataGridPremium component to unlock this feature.',
+          ].join('\n'),
+        );
+      }
+
+      if (newRows.length && newRows.length !== lastRowToRender - firstRowToRender) {
+        throw new Error(
+          [
+            'MUI: The number of rows you want update is more than what was indicated',
```

Comment:
It could be less as well
```suggestion
            'MUI: The number of rows you want update is different than what was indicated',
```